### PR TITLE
[Diagnostics] Emit a diagnostic if a function parameter is unused in the body

### PR DESCRIPTION
**Pull request by theblixguy**
When a function parameter is not used in the body of the function, emit a warning. For example:

```swift
func foo(x: Int) { // warning + fix-it to replace 'x' with '_' OR insert '_' after 'x'
  print("I didn't use x!")
}

func bar(_ y: Int) { // warning + fix-it remove 'y'
  print("I didn't use y!")
}

func baz(x y: Int) { // warning + fix-it to replace entire parameter with '_' OR replace 'y' with '_'
  print("I didn't use y!")
}
```

This is hidden behind a frontend flag: `-warn-unused-function-params`. When this flag is passed, we emit the warnings, otherwise we don't. We probably also need a new Xcode build setting to conveniently allow the users to enable or disable this functionality, but that's not up to me.

In the future, we can extend this to work with closure parameters as well.

Resolves SR-2849
Resolves SR-6963
Resolves rdar://problem/31844597
Resolves rdar://problem/35534022

### CODE REVIEW

**Comment by theblixguy** [comment]
This is just my first pass - I may have missed some edge cases or perhaps we need to handle some of these scenarios a little differently.

**Comment by owenv** [comment]
I'm hesitant to start adding user-visible flags to enable/disable warnings, this would be the first one not related to resilience or language modes afaik. I think it deserves a discussion either around unconditionally enabling this warning or changing Swift's warnings policy and adding new warning control features on a broader scale. 

In the meantime, maybe this could be a help-hidden no-driver option?

**Comment by theblixguy** [comment]
We already have similar flags (`warn_implicit_overrides` & `warn-swift3-objc-inference-{minimal|complete}`. This flag just mirrors `GCC_WARN_UNUSED_PARAMETER`. There was a bit of discussion on it on the forums in the past: https://forums.swift.org/t/swift-and-gcc-warn-unused-parameter/14607/5

**Comment by owenv** [comment]
I'd consider `warn-swift3-objc-inference` a historical oddity, but I wasn't aware of `warn-implicit-overrides`. I'm still kind of opposed to adding to the list of configurable warnings on an ad hoc basis, but I'll leave the final decision to others.

File: test/decl/func/warn_unused_function_params.swift
Comment (LucianoPAlmeida) [inline test/decl/func/warn_unused_function_params.swift]: Just an opinion here is that `consider replacing it with '_'` to me doesn't see the ideal suggestion and fix-it. Because if you have a parameter in a function that you don't use, the preferable fix would be to remove it. I'm not sure having a function `func foo6(x _: Int)` where a user is forced to pass a parameter that the function wouldn't even use it will be a ideal fix. Just pointing this question out to know if it makes sense and what you guys think :))

Comment (theblixguy) [reply test/decl/func/warn_unused_function_params.swift]: We could offer another fix-it to remove it entirely

Comment (theblixguy) [reply test/decl/func/warn_unused_function_params.swift]: I have decided to offer both fix-its. For example, in the original bug report, there was an example of `UITableView` method where ignoring just the parameter name would be the right thing to do, instead of completely ignoring the parameter.

Comment (LucianoPAlmeida) [reply test/decl/func/warn_unused_function_params.swift]: Humm I see. Makes sense because the function signature has to match the protocol requirements :)

**Review by beccadax — CHANGES_REQUESTED** [review]
On a technical level:

* This warning would apply equally well to initializers and subscripts. It looks like you *might* actually cover those, but you’re certainly not testing them.
* This warning should not apply to parameters which are mentioned in a `- Parameter` field in their doc comment. This probably won’t appear as a use in the AST, but it’s still a “use” in that removing the parameter name “breaks” the documentation. I’m not sure if there’s a practical way to detect this, however.
* This warning should probably not apply to public or open declarations. Internal names are visible to clients of a library and serve as soft documentation.
* I don’t think it makes sense to offer a fix-it that removes the argument label—applying it would break call sites. Even if you did include that feature, you would absolutely need to make sure the member doesn’t witness a protocol requirement, override a superclass method, or get overridden by a subclass. And you would need to make sure that, when there are two possible options to fix the diagnostic, they are both attached to separate notes, instead of one of them being attached to a note and the other to the warning itself.

-----

However, as a matter of policy, I don’t think we should merge this PR even if these issues are addressed.

Swift typically only includes warnings that we believe are so likely to be correct and helpful that we feel confident enabling them unconditionally. There are a few exceptions, but they tend to relate to deprecated features or things that are very difficult to detect without running the full compiler. Unused parameters are not one of those things—linters are pretty good at detecting this condition, and there are a number of reasons why you might want to leave the parameter names in place (such as to help people understand the intended purpose of the parameter).

If you want to get this change in, I think you should start a discussion and gain broad consensus that we would like to include these kinds of warnings. Well, either that or get someone to convince me that I’m wrong. 😉

**Comment by theblixguy** [comment]
Sure, we can discuss this further in the forums.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**
